velox/functions/lib/Re2Functions.h

```diff
@@ -114,7 +116,68 @@ std::vector<std::shared_ptr<exec::FunctionSignature>> re2ExtractSignatures();
 /// prefix, and suffix patterns. Return the pair {pattern kind, number of '_'
 /// characters} for patterns with wildcard characters only. Return
 /// {kGenericPattern, 0} for generic patterns).
-PatternMetadata determinePatternKind(StringView pattern);
+PatternMetadata determinePatternKind(
+    StringView pattern,
+    std::optional<char> escapeChar);
+
+/// Return the unescaped string for the specified string range, if escape char
+/// is not specified just return the corresponding substring.
+std::string unescape(
+    StringView pattern,
+    vector_size_t start,
+    vector_size_t end,
+    std::optional<char> escapeChar);
+std::string unescape(StringView pattern, std::optional<char> escapeChar);
+
+/// An Iterator that provides methods(hasNext, next) to iterate through a
+/// pattern string.
+class PatternStringIterator {
+ public:
+  struct State {
+    // Is current char the escape char?
+    // NOTE: If escape char is set as '\',  for pattern '\\', the first '\' is
+    // an escaping char, the second is not, it is just a literal '\'
+    bool isEscaping = false;
+    // Is current char the wildcard char?
+    // NOTE: If escape char is set as '\', for pattern '\%%', the first '%' is
+    // not a wildcard, just a literal '%', the second '%' is a wildcard.
+    bool isWildcard = false;
+  };
+
+  explicit PatternStringIterator(
+      StringView pattern,
+      std::optional<char> escapeChar)
+      : pattern_(pattern), escapeChar_(escapeChar) {}
+
+  bool hasNext();
```

Comment:
please, document all these methods and mark some as const; it is hard to understand what these do